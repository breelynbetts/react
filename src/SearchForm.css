.SearchForm {
  grid-area: logo;
  display: inline;
  flex-direction: column;
  margin-top: 0;
}

.SearchForm > .ButtonBar {
  display: inline;
}

.SearchForm > .ButtonBar > button {
  margin-left: auto;
  margin-top: 1rem;
}

.SearchForm .error {
  background-color: pink;
  border: solid 1px red;
  color: maroon;
  margin: 1rem;
  padding: 1rem;
}
