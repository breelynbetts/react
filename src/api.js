const apiHost = () => {} // No-op in our mock version.

const searchNasa = () => Promise.resolve({
  data: [{
      "id": 689597,
      "sol": 2540,
      "camera": {
        "id": 20,
        "name": "FHAZ",
        "rover_id": 5,
        "full_name": "Front Hazard Avoidance Camera"
      },
      "img_src": "https://mars.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/02540/opgs/edr/fcam/FLB_622990123EDR_F0763002FHAZ00341M_.JPG",
      "earth_date": "2019-09-28",
      "rover": {
        "id": 5,
        "name": "Curiosity",
        "landing_date": "2012-08-06",
        "launch_date": "2011-11-26",
        "status": "active",
        "max_sol": 2540,
        "max_date": "2019-09-28",
        "total_photos": 366206,
        "cameras": [{
            "name": "FHAZ",
            "full_name": "Front Hazard Avoidance Camera"
          },
          {
            "name": "NAVCAM",
            "full_name": "Navigation Camera"
          },
          {
            "name": "MAST",
            "full_name": "Mast Camera"
          },
          {
            "name": "CHEMCAM",
            "full_name": "Chemistry and Camera Complex"
          },
          {
            "name": "MAHLI",
            "full_name": "Mars Hand Lens Imager"
          },
          {
            "name": "MARDI",
            "full_name": "Mars Descent Imager"
          },
          {
            "name": "RHAZ",
            "full_name": "Rear Hazard Avoidance Camera"
          }
        ]
      }
    },
    {
      "id": 689598,
      "sol": 2540,
      "camera": {
        "id": 20,
        "name": "FHAZ",
        "rover_id": 5,
        "full_name": "Front Hazard Avoidance Camera"
      },
      "img_src": "https://mars.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/02540/opgs/edr/fcam/FRB_622990123EDR_F0763002FHAZ00341M_.JPG",
      "earth_date": "2019-09-28",
      "rover": {
        "id": 5,
        "name": "Curiosity",
        "landing_date": "2012-08-06",
        "launch_date": "2011-11-26",
        "status": "active",
        "max_sol": 2540,
        "max_date": "2019-09-28",
        "total_photos": 366206,
        "cameras": [{
            "name": "FHAZ",
            "full_name": "Front Hazard Avoidance Camera"
          },
          {
            "name": "NAVCAM",
            "full_name": "Navigation Camera"
          },
          {
            "name": "MAST",
            "full_name": "Mast Camera"
          },
          {
            "name": "CHEMCAM",
            "full_name": "Chemistry and Camera Complex"
          },
          {
            "name": "MAHLI",
            "full_name": "Mars Hand Lens Imager"
          },
          {
            "name": "MARDI",
            "full_name": "Mars Descent Imager"
          },
          {
            "name": "RHAZ",
            "full_name": "Rear Hazard Avoidance Camera"
          }
        ]
      }
    },
    {
      "id": 689599,
      "sol": 2540,
      "camera": {
        "id": 21,
        "name": "RHAZ",
        "rover_id": 5,
        "full_name": "Rear Hazard Avoidance Camera"
      },
      "img_src": "https://mars.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/02540/opgs/edr/rcam/RLB_622990164EDR_F0763002RHAZ00341M_.JPG",
      "earth_date": "2019-09-28",
      "rover": {
        "id": 5,
        "name": "Curiosity",
        "landing_date": "2012-08-06",
        "launch_date": "2011-11-26",
        "status": "active",
        "max_sol": 2540,
        "max_date": "2019-09-28",
        "total_photos": 366206,
        "cameras": [{
            "name": "FHAZ",
            "full_name": "Front Hazard Avoidance Camera"
          },
          {
            "name": "NAVCAM",
            "full_name": "Navigation Camera"
          },
          {
            "name": "MAST",
            "full_name": "Mast Camera"
          },
          {
            "name": "CHEMCAM",
            "full_name": "Chemistry and Camera Complex"
          },
          {
            "name": "MAHLI",
            "full_name": "Mars Hand Lens Imager"
          },
          {
            "name": "MARDI",
            "full_name": "Mars Descent Imager"
          },
          {
            "name": "RHAZ",
            "full_name": "Rear Hazard Avoidance Camera"
          }
        ]
      }
    }
  ]
})

const searchAstronomyPicture = () => Promise.resolve({
  data: [{
    "copyright": "Eric Wagner",
    "date": "2019-10-10",
    "explanation": "On September 24, a late evening commercial flight from Singapore to Australia offered stratospheric views of the southern hemisphere's night sky, if you chose a window seat. In fact, a well-planned seating choice with a window facing toward the Milky Way allowed the set up of a sensitive digital camera on a tripod mount to record the galaxy's central bulge in a series of 10 second long exposures.  By chance, one of the exposures caught this bright fireball meteor in the starry frame. Reflected along the wing of the A380 aircraft, the brilliant greenish streak is also internally reflected in the double layer window, producing a fainter parallel to the original meteor track. In the southern sky Jupiter is the bright source beneath the galactic bulge and seen next to a green beacon, just off the wing tip.",
    "hdurl": "https://apod.nasa.gov/apod/image/1910/MWBolideEricWagner2400.jpg",
    "media_type": "image",
    "service_version": "v1",
    "title": "Mid-Air Meteor and Milky Way",
    "url": "https://apod.nasa.gov/apod/image/1910/MWBolideEricWagner1200.jpg"
  }]

})

const searchEvents = () => Promise.resolve({
  data: [{
    "title": "EONET Events",
    "description": "Natural events from EONET.",
    "link": "https://eonet.sci.gsfc.nasa.gov/api/v2.1/events",
    "events": [
      {
        "id": "EONET_4475",
        "title": "Wildfires - Los Angeles County (Tick Fire), California, United States",
        "description": "",
        "link": "https://eonet.sci.gsfc.nasa.gov/api/v2.1/events/EONET_4475",
        "categories": [
          {
            "id": 8,
            "title": "Wildfires"
          }
        ],
        "sources": [
          {
            "id": "PDC",
            "url": "http://emops.pdc.org/emops/?hazard_id=96981"
          }
        ],
        "geometries": [
          {
            "date": "2019-10-24T18:16:00Z",
            "type": "Point",
            "coordinates": [
              -118.39468413,
              34.457634347
            ]
          }
        ]
      },
      {
        "id": "EONET_4476",
        "title": "Wildfires - San Bernardino County (Old Water Fire), California, United States",
        "description": "",
        "link": "https://eonet.sci.gsfc.nasa.gov/api/v2.1/events/EONET_4476",
        "categories": [
          {
            "id": 8,
            "title": "Wildfires"
          }
        ],
        "sources": [
          {
            "id": "PDC",
            "url": "http://emops.pdc.org/emops/?hazard_id=96977"
          }
        ],
        "geometries": [
          {
            "date": "2019-10-24T15:11:00Z",
            "type": "Point",
            "coordinates": [
              -117.26913,
              34.17922
            ]
          }
        ]
      },
      {
        "id": "EONET_4474",
        "title": "Wildfire -Sonoma County (Kincade Fire), California, United States",
        "description": "",
        "link": "https://eonet.sci.gsfc.nasa.gov/api/v2.1/events/EONET_4474",
        "categories": [
          {
            "id": 8,
            "title": "Wildfires"
          }
        ],
        "sources": [
          {
            "id": "PDC",
            "url": "http://emops.pdc.org/emops/?hazard_id=96971"
          }
        ],
        "geometries": [
          {
            "date": "2019-10-24T13:32:00Z",
            "type": "Point",
            "coordinates": [
              -122.807158563,
              38.758962281
            ]
          }
        ]
      }
    ]
  }  
  ]

})

export {
  apiHost,
  searchNasa,
  searchAstronomyPicture,
  searchEvents
}
