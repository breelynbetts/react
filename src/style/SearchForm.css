@import url('https://fonts.googleapis.com/css?family=Fjalla+One|Oswald&display=swap');

.SearchForm > .ButtonBar {
  display: inline;
}

.SearchForm {
  height: 25px;
  align-items: center;
}

.SearchForm input {
  width: 30%;
  height: 80%;
}

.ButtonBar button:disabled {
  background-color: grey;  
}

.ButtonBar button {
  background-color: #b4c1ff;
  height: 100%; 
  width: 100px;
  font-family: 'Fjalla One', sans-serif;
  letter-spacing: 1px;
  font-size: 13.375px;
  letter-spacing: 2px;
}

