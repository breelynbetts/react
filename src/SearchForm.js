import React, { useState } from 'react'

import './style/SearchForm.css'

import SearchResults from './SearchResults'

import { searchNasa } from './api'

const SearchForm = () => {
  const [error, setError] = useState(null)
  const [query, setQuery] = useState('')
  const [images, setImages] = useState([])

  const handleQueryChange = event => setQuery(event.target.value)

  const performQuery = async event => {
    event.preventDefault()

    setError(null)

    try {
      const result = await searchNasa({
        q: query
      })

      setImages(result.data)
    } catch (error) {
      setError('Sorry, but something went wrong.')
    }
  }
  
  return (
    <div>
      <form className="SearchForm" onSubmit={performQuery}>

        <input name="query" type="text" value={query} onChange={handleQueryChange} />

        <div className="ButtonBar">
          <button type="submit" disabled={!query}>Search</button>
        </div>

        {error && (
          <div className="error">
            {error}
            <p>Error has occurred!</p>
          </div>
        )}
      </form>
      <div className='ImageResults'>
        <SearchResults results={images} />
      </div>
    
    </div>
  )
}

export default SearchForm
