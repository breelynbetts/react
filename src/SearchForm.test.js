import React from 'react'
import ReactDOM from 'react-dom'
import TestRenderer from 'react-test-renderer'
import ReactTestUtils from 'react-dom/test-utils'

import sinon from 'sinon'

import SearchForm from './SearchForm'

import * as api from './api'


it('should start with an empty search field', () => {
  const component = TestRenderer.create(<SearchForm />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

it('should start with a disabled search button', () => {
  const component = TestRenderer.create(<SearchForm />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('search button', () => {
  let div
  beforeEach(() => {
    div = document.createElement('div')
    ReactTestUtils.act(() => {
      ReactDOM.render(<SearchForm />, div)
    })
  })

  afterEach(() => ReactDOM.unmountComponentAtNode(div))

  it('should be enabled when the search field is not blank', () => {
    const searchInput = div.querySelector('input')
    ReactTestUtils.act(() => {
      searchInput.value = 'i can haz unit tests'
      ReactTestUtils.Simulate.change(searchInput)
    })

    const searchButton = div.querySelector('button')
    expect(searchButton.disabled).toBe(false)
  })

  it('should be disabled when the search field is blank', () => {
    const searchInput = div.querySelector('input')
    ReactTestUtils.act(() => {
      searchInput.value = ''
      ReactTestUtils.Simulate.change(searchInput)
    })

    const searchButton = div.querySelector('button')
    expect(searchButton.disabled).toBe(true)
  })
})

// Helper function for the next two test collections.
const setupAndQuerySearchForm = async () => {
  const div = document.createElement('div')
  ReactTestUtils.act(() => {
    ReactDOM.render(<SearchForm />, div)
  })

  const searchInput = div.querySelector('input')
  ReactTestUtils.act(() => {
    searchInput.value = 'hello'
    ReactTestUtils.Simulate.change(searchInput)
  })

  const searchForm = div.querySelector('form')
  await ReactTestUtils.act(async () => {
    await ReactTestUtils.Simulate.submit(searchForm)
  })

  return div
}

describe('API calls', () => {
  let div
  beforeEach(async () => {
    sinon.stub(api, 'searchNasa')

    // To manage size, we supply a mock response that contains _only_ what the app will need. This does mean
    // that we need to revise the mock response if our app starts using more (or different) data.
    api.searchNasa.returns(Promise.resolve({
      collection: 
        {
          "metadata": {
            "total_hits": 596
          },
          "version": "1.0",
          "items": [{
            "data": [
              {
                "description_508": "NASA's GALEX reveals the true nature of UGC 1382, dubbed the Frankenstein galaxy. Scientists have discovered that UGC 1382 is a giant, and one of the largest isolated galaxies known.",
                "center": "JPL",
                "nasa_id": "PIA20695",
                "description": "The galaxy UGC 1382 has been revealed to be far larger and stranger than previously thought. Astronomers relied on a combination of ground-based and space telescopes to uncover the true nature of this \"Frankenstein galaxy.\" The composite image shows the same galaxy as viewed with different instruments. The component images are also available.  In the image at left, UGC 1382 appears to be a simple elliptical galaxy, based on optical data from the Sloan Digital Sky Survey (SDSS). But spiral arms emerged when astronomers incorporated ultraviolet data from the Galaxy Evolution Explorer (GALEX) and deep optical data from SDSS, as seen in the middle image. Combining that with a view of low-density hydrogen gas (shown in green), detected at radio wavelengths by the Very Large Array, scientists discovered that UGC 1382 is a giant, and one of the largest isolated galaxies known.  GALEX in particular was able detect very faint features because it operated from space, which is necessary for UV observations because ultraviolet light is absorbed by the Earth's atmosphere. Astronomers also used Stripe 82 of SDSS, a small region of sky where SDSS imaged the sky 80 times longer than the original standard SDSS survey. This enabled optical detection of much fainter features as well.  http://photojournal.jpl.nasa.gov/catalog/PIA20695",
                "secondary_creator": "NASA/JPL/Caltech/SDSS/NRAO",
                "date_created": "2016-07-11T15:50:52Z",
                "keywords": [
                  "Galaxy Evolution Explorer GALEX"
                ],
                "title": "Frankenstein Galaxy",
                "media_type": "image"
              }
            ],
            "links": [{
              "render": "image",
              "rel": "preview",
              "href": "https://images-assets.nasa.gov/image/PIA20695/PIA20695~thumb.jpg"
            }],
            "href": "https://images-assets.nasa.gov/image/PIA20695/collection.json"
          }]
      }
    }))

    div = await setupAndQuerySearchForm()
  })

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(div)
    api.searchNasa.restore()
  })

  it('should trigger a NASA search when the search button is clicked', () => {
    // Note how this _isn’t_ a snapshot test because we’re checking whether a function was called with
    // the right arguments.
    expect(api.searchNasa.firstCall.args[0]).toEqual({
      q: 'hello',
      media_type: "image", // Our test search term.
    })
  })

  it('should populate the image container when search results arrive', () => {
    // Our mock search results yield one image, so we expect our results container to have one child.
    const searchResults = div.querySelector('div.ImageResults')
    expect(searchResults.children.length).toEqual(1)
  })
})

describe('failed API calls', () => {
  let div
  beforeEach(async () => {
    sinon.stub(api, 'searchNasa')
    api.searchNasa.returns(Promise.reject('Mock failure'))

    div = await setupAndQuerySearchForm()
  })

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(div)
    api.searchNasa.restore()
  })

  it('should display an alert when the API call fails', () => {
    // The document should contain the error div.
    const searchError = div.querySelector('div.error')
    expect(searchError.textContent).toEqual('Sorry, but something went wrong.')
  })
})