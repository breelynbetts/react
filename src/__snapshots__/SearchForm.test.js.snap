// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should start with a disabled search button 1`] = `
<div>
  <form
    className="SearchForm"
    onSubmit={[Function]}
  >
    <input
      name="query"
      onChange={[Function]}
      type="text"
      value=""
    />
    <div
      className="ButtonBar"
    >
      <button
        disabled={true}
        type="submit"
      >
        search
      </button>
    </div>
  </form>
  <div
    className="ImageResults"
  >
    <div
      className="search"
    />
  </div>
</div>
`;

exports[`should start with an empty search field 1`] = `
<div>
  <form
    className="SearchForm"
    onSubmit={[Function]}
  >
    <input
      name="query"
      onChange={[Function]}
      type="text"
      value=""
    />
    <div
      className="ButtonBar"
    >
      <button
        disabled={true}
        type="submit"
      >
        search
      </button>
    </div>
  </form>
  <div
    className="ImageResults"
  >
    <div
      className="search"
    />
  </div>
</div>
`;
